stage0/src/zero_page.rs

```diff
@@ -59,6 +79,41 @@ pub fn init_zero_page(fw_cfg: &mut FwCfg) -> &'static mut BootParams {
     zero_page
 }
 
+/// Builds an E820 table by reading the low and high memory amount from CMOS.
+///
+/// The code is largely based on what SeaBIOS is doing (see `qemu_preinit()` and `qemu_cfg_e820()`
+/// in <https://github.com/qemu/seabios/blob/b0d61ecef66eb05bd7a4eb7ada88ec5dab06dfee/src/fw/paravirt.c>),
+/// but <https://wiki.osdev.org/Detecting_Memory_%28x86%29> is also a good read on the topic.
+fn build_e820_from_nvram(
+    zero_page: &mut BootParams,
+    port_factory: PortFactoryWrapper,
+) -> Result<(), &'static str> {
+    // Safety: (a) fw_cfg is available, so we're running under QEMU(ish) and (b) there was no
+    // pre-built E820 table in fw_cfg; thus, we can reasonably expect CMOS to available, as that's
+    // what SeaBIOS would use in that situation to build the E820 table.
+    let mut cmos = unsafe { Cmos::new(port_factory) };
+    let mut rs = cmos.low_ram_size()?;
+    let high = cmos.high_ram_size()?;
+
+    // Time to put all that we know together.
+    // Leave the last 4 pages of low memory as reserved: this includes the BIOS area, and other
+    // things that according to SeaBIOS are reserved.
+    if rs > 0xFFFB_C000 {
+        rs = 0xFFFB_C000;
+    };
+    zero_page.e820_entries = 2;
+    zero_page.e820_table[0] = BootE820Entry::new(0, rs as usize, E820EntryType::RAM);
+    zero_page.e820_table[1] = BootE820Entry::new(0xFFFB_C000, 4 * 0x1000, E820EntryType::RESERVED);
```

Comment:
From looking at the SEABIOS implementation, it seems that it also adds the 256KiB below 4GiB as reserved for the BIOS image (everything above 0xFFFC_0000).

Reply:
Ah, good catch, I somehow lost it in the clean-ups!

My idea here was that as [0xFFFB_C000, 0xFFFC_0000) is just before [0xFFFC_0000, 0x1_0000_0000), we can just reserve all of it with one entry. But I somehow lost the reservation for the higher 256K.